taichi/system/unified_allocator.cpp

```diff
@@ -30,8 +32,16 @@ UnifiedAllocator::UnifiedAllocator(std::size_t size, Arch arch)
     // This could be run on a host worker thread, so we have to set the context
     // before using any of the CUDA driver function call.
     auto _ = CUDAContext::get_instance().get_guard();
-    CUDADriver::get_instance().malloc_managed(&_cuda_data, size,
-                                              CU_MEM_ATTACH_GLOBAL);
+
+    Device::AllocParams alloc_params;
+    alloc_params.size = size;
+    alloc_params.host_read = true;
+    alloc_params.host_write = true;
+
+    cuda::CudaDevice *cuda_device = static_cast<cuda::CudaDevice *>(device);
```

Comment:
DeviceAllocation carries a ptr to Device as well. Random thought: can we delay initializing device_(device_ = std::unique_ptr<Device>(new cuda::CudaDevice());) until here instead of passing it all the way down from LlvmProgramImpl?

Reply:
Hmm I'm not quite following, are you suggesting that we let `unified_allocator` own the `device`? Also I imagine in the future the `device_` will be used in many other places, so very soon this might not be its first usage anymore. I find that it makes more sense to initialize it within `ProgramImpl`.

Reply:
Yea I was indeed thinking about simplifying the ownership and trying to avoid passing around a device ptr. ;)  But if our desired end state is to let Device own the memoryPool this makes sense to me!
(nit: IMHO an easier process is creating a `DesiredDevice` class and make it work e2e, then rename & remove the old Device class in the end to show people the difference before and after the refactor ;)

Reply:
What would be the difference between `DesiredDevice` and `Device`?

Reply:
I'd imagine we first create a `DesiredDevice` to include the MemoryPool, keep adding it to the codebase until it works e2e. And finally remove `Device` class and rename `DesiredDevice` to `Device`. ;) The main advantage is we always have the old and new classes for reviewers to compare even in the intermediate PRs. But ofc this is my personal choice to make smaller PRs haha.  It's totally up to you how to split the refactor into PRs ;)

Reply:
I see! That makes sense 🤔